From 053c6a433d38cd1e1176fb5aca27de98f12c290a Mon Sep 17 00:00:00 2001
From: Yafen Fang <yafen@iscas.ac.cn>
Date: Wed, 1 Apr 2026 18:16:22 +0800
Subject: [PATCH] add support for openRuyi

Add openRuyi Product to provide security configuration baseline scanning capability for openRuyi.
1. Build the openRuyi-related support framework
2. Add common openRuyi check items
---
 CMakeLists.txt                                |   5 +
 build_product                                 |   1 +
 components/filesystem.yml                     |   2 +
 controls/std_openruyi.yml                     | 415 ++++++++++++++++++
 .../obsolete/service_rsyncd_disabled/rule.yml |   1 +
 .../guide/services/ssh/sshd_strong_kex.var    |   1 +
 .../oval/shared.xml                           |   2 +-
 .../require_emergency_target_auth/rule.yml    |   2 +-
 .../tests/invalid_username.fail.sh            |   2 +-
 .../oval/shared.xml                           |  34 +-
 .../files/file_empty_link_prohibit/rule.yml   |  26 ++
 .../file_empty_link_prohibit/sce/shared.sh    |  11 +
 .../file_hidden_executable_prohibit/rule.yml  |  25 ++
 .../sce/shared.sh                             |  13 +
 product_properties/10-grub.yml                |   5 +
 products/openruyi/CMakeLists.txt              |   6 +
 products/openruyi/product.yml                 |  24 +
 products/openruyi/profiles/standard.profile   |  14 +
 shared/applicability/bootc.yml                |   2 +
 shared/applicability/oval/bootc.xml           |   2 +
 .../oval/installed_OS_is_openruyi_riscv.xml   |  25 ++
 .../oval/installed_env_has_grub2_package.xml  |   4 +-
 ...proc_sys_kernel_osrelease_arch_riscv64.xml |   1 +
 .../applicability/oval/system_with_kernel.xml |   2 +
 shared/applicability/package.yml              |   4 +-
 shared/applicability/riscv64_arch.yml         |   5 +
 shared/applicability/system_with_kernel.yml   |   4 +
 shared/checks/oval/bootc.xml                  |   2 +
 .../checks/oval/installed_OS_is_openruyi.xml  |  24 +
 .../oval/sysctl_kernel_ipv6_disable.xml       |   1 +
 .../oval/system_info_architecture_64bit.xml   |   2 +
 .../system_info_architecture_riscv_64.xml     |  29 ++
 shared/macros/01-general.jinja                |   2 +
 shared/macros/10-ansible.jinja                |   6 +
 shared/macros/10-bash.jinja                   |  10 +-
 shared/macros/10-oval.jinja                   |   2 +-
 .../templates/accounts_password/bash.template |   4 +-
 .../templates/accounts_password/oval.template |   6 +-
 .../grub2_bootloader_argument/bash.template   |   2 +-
 .../bash.template                             |   2 +-
 .../templates/service_disabled/bash.template  |   2 +-
 .../templates/service_enabled/bash.template   |   2 +-
 shared/templates/sysctl/bash.template         |   2 +-
 ssg/constants.py                              |   6 +-
 tests/shared/grub2.sh                         |   4 +-
 45 files changed, 721 insertions(+), 25 deletions(-)
 create mode 100644 controls/std_openruyi.yml
 create mode 100644 linux_os/guide/system/permissions/files/file_empty_link_prohibit/rule.yml
 create mode 100644 linux_os/guide/system/permissions/files/file_empty_link_prohibit/sce/shared.sh
 create mode 100644 linux_os/guide/system/permissions/files/file_hidden_executable_prohibit/rule.yml
 create mode 100644 linux_os/guide/system/permissions/files/file_hidden_executable_prohibit/sce/shared.sh
 create mode 100644 products/openruyi/CMakeLists.txt
 create mode 100644 products/openruyi/product.yml
 create mode 100644 products/openruyi/profiles/standard.profile
 create mode 100644 shared/applicability/oval/installed_OS_is_openruyi_riscv.xml
 create mode 100644 shared/applicability/oval/proc_sys_kernel_osrelease_arch_riscv64.xml
 create mode 100644 shared/applicability/riscv64_arch.yml
 create mode 100644 shared/checks/oval/installed_OS_is_openruyi.xml
 create mode 100644 shared/checks/oval/system_info_architecture_riscv_64.xml

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 48b206911b..a62ff25915 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -104,6 +104,7 @@ option(SSG_PRODUCT_OL9 "If enabled, the Oracle Linux 9 SCAP content will be buil
 option(SSG_PRODUCT_OL10 "If enabled, the Oracle Linux 10 SCAP content will be built" ${SSG_PRODUCT_DEFAULT})
 option(SSG_PRODUCT_OPENEMBEDDED "If enabled, the OpenEmbedded SCAP content will be built" ${SSG_PRODUCT_DEFAULT})
 option(SSG_PRODUCT_OPENEULER2203 "If enabled, the openEuler 22.03 LTS content will be built" ${SSG_PRODUCT_DEFAULT})
+option(SSG_PRODUCT_OPENRUYI "If enabled, the openRuyi SCAP content will be built" ${SSG_PRODUCT_DEFAULT})
 option(SSG_PRODUCT_OPENSUSE "If enabled, the openSUSE SCAP content will be built" ${SSG_PRODUCT_DEFAULT})
 option(SSG_PRODUCT_RHCOS4 "If enabled, the RHCOS4 SCAP content will be built" ${SSG_PRODUCT_DEFAULT})
 option(SSG_PRODUCT_RHEL8 "If enabled, the RHEL8 SCAP content will be built" ${SSG_PRODUCT_DEFAULT})
@@ -342,6 +343,7 @@ message(STATUS "Oracle Linux 9: ${SSG_PRODUCT_OL9}")
 message(STATUS "Oracle Linux 10: ${SSG_PRODUCT_OL10}")
 message(STATUS "OpenEmbedded: ${SSG_PRODUCT_OPENEMBEDDED}")
 message(STATUS "openEuler 22.03 LTS: ${SSG_PRODUCT_OPENEULER2203}")
+message(STATUS "openRuyi: ${SSG_PRODUCT_OPENRUYI}")
 message(STATUS "openSUSE: ${SSG_PRODUCT_OPENSUSE}")
 message(STATUS "RHEL 8: ${SSG_PRODUCT_RHEL8}")
 message(STATUS "RHEL 9: ${SSG_PRODUCT_RHEL9}")
@@ -445,6 +447,9 @@ endif()
 if(SSG_PRODUCT_OPENEULER2203)
     add_subdirectory("products/openeuler2203" "openeuler2203")
 endif()
+if(SSG_PRODUCT_OPENRUYI)
+    add_subdirectory("products/openruyi" "openruyi")
+endif()
 if(SSG_PRODUCT_OPENSUSE)
     add_subdirectory("products/opensuse" "opensuse")
 endif()
diff --git a/build_product b/build_product
index 76e3d3a698..49b1a9fd4f 100755
--- a/build_product
+++ b/build_product
@@ -348,6 +348,7 @@ all_cmake_products=(
 	OL10
 	OPENEMBEDDED
 	OPENEULER2203
+	OPENRUYI
 	OPENSUSE
 	RHCOS4
 	RHEL8
diff --git a/components/filesystem.yml b/components/filesystem.yml
index bae7bd640e..ca95148629 100644
--- a/components/filesystem.yml
+++ b/components/filesystem.yml
@@ -197,6 +197,8 @@ rules:
 - sysctl_fs_protected_hardlinks
 - sysctl_fs_protected_regular
 - sysctl_fs_protected_symlinks
+- file_empty_link_prohibit
+- file_hidden_executable_prohibit
 templates:
 - mount
 - mount_option
diff --git a/controls/std_openruyi.yml b/controls/std_openruyi.yml
new file mode 100644
index 0000000000..ac707d41ec
--- /dev/null
+++ b/controls/std_openruyi.yml
@@ -0,0 +1,415 @@
+---
+policy: 'Standard Benchmark for openRuyi'
+title: 'Standard Benchmark for openRuyi'
+id: std_openruyi
+version: '1.0'
+levels:
+  - id: base
+
+controls:
+  - id: 1.1_bootloader_password
+    title: Ensure Bootloader Password is Set
+    levels:
+      - base
+    status: automated
+    rules:
+      - grub2_password
+      - grub2_password.severity=high
+      - grub2_uefi_password
+      - grub2_uefi_password.severity=high
+
+  - id: 1.2_single_user_mode_password
+    title: Ensure Single User Mode Requires Authentication
+    levels:
+      - base
+    status: automated
+    rules:
+      - require_singleuser_auth
+      - require_singleuser_auth.severity=high
+      - require_emergency_target_auth
+      - require_emergency_target_auth.severity=high
+
+  - id: 1.3_umask_configured
+    title: Ensure Umask is Configured Correctly
+    levels:
+      - base
+    status: automated
+    rules:
+      - accounts_umask_etc_bashrc
+      - accounts_umask_etc_bashrc.severity=high
+      - accounts_umask_etc_login_defs
+      - accounts_umask_etc_login_defs.severity=high
+      - accounts_umask_etc_profile
+      - accounts_umask_etc_profile.severity=high
+      - var_accounts_user_umask=027
+
+  - id: 1.4_filesystem_permissions
+    title: Ensure Filesystem Permissions are Secure
+    levels:
+      - base
+    status: automated
+    rules:
+      - dir_perms_world_writable_sticky_bits
+      - dir_perms_world_writable_sticky_bits.severity=high
+      - file_permissions_unauthorized_world_writable
+      - file_permissions_unauthorized_world_writable.severity=high
+      - no_files_unowned_by_user
+      - no_files_unowned_by_user.severity=high
+      - file_permissions_ungroupowned
+      - file_permissions_ungroupowned.severity=high
+      - file_empty_link_prohibit
+      - file_empty_link_prohibit.severity=high
+      - file_hidden_executable_prohibit
+      - file_hidden_executable_prohibit.severity=high
+
+  - id: 1.5_critical_files_ownership
+    title: Ensure Critical Files Ownership is Correct
+    levels:
+      - base
+    status: automated
+    rules:
+      - file_owner_etc_passwd
+      - file_owner_etc_passwd.severity=high
+      - file_groupowner_etc_passwd
+      - file_groupowner_etc_passwd.severity=high
+      - file_permissions_etc_passwd
+      - file_permissions_etc_passwd.severity=high
+      - file_owner_etc_shadow
+      - file_owner_etc_shadow.severity=high
+      - file_groupowner_etc_shadow
+      - file_groupowner_etc_shadow.severity=high
+      - file_permissions_etc_shadow
+      - file_permissions_etc_shadow.severity=high
+      - file_owner_etc_group
+      - file_owner_etc_group.severity=high
+      - file_groupowner_etc_group
+      - file_groupowner_etc_group.severity=high
+      - file_permissions_etc_group
+      - file_permissions_etc_group.severity=high
+      - file_owner_etc_gshadow
+      - file_owner_etc_gshadow.severity=high
+      - file_groupowner_etc_gshadow
+      - file_groupowner_etc_gshadow.severity=high
+      - file_permissions_etc_gshadow
+      - file_permissions_etc_gshadow.severity=high
+
+  - id: 1.6_crypto_policy_configured
+    title: Ensure System Crypto Policy is Configured
+    levels:
+      - base
+    status: automated
+    rules:
+      - configure_crypto_policy
+      - configure_crypto_policy.severity=high
+      - var_system_crypto_policy=default_policy
+
+  - id: 2.1_auditd_configured
+    title: Ensure Auditd is Configured
+    levels:
+      - base
+    status: automated
+    rules:
+      - package_audit_installed
+      - package_audit_installed.severity=high
+      - service_auditd_enabled
+      - service_auditd_enabled.severity=high
+      - grub2_audit_argument
+      - grub2_audit_argument.severity=low
+      - grub2_audit_backlog_limit_argument
+      - grub2_audit_backlog_limit_argument.severity=low
+      - var_auditd_max_log_file=6
+      - var_auditd_max_log_file_action=rotate
+      - var_audit_backlog_limit=8192
+
+  - id: 2.2_audit_rules_configured
+    title: Ensure Critical Audit Rules are Loaded
+    levels:
+      - base
+    status: automated
+    rules:
+      - audit_rules_immutable
+      - audit_rules_immutable.severity=high
+      - audit_rules_sysadmin_actions
+      - audit_rules_sysadmin_actions.severity=high
+      - audit_rules_suid_privilege_function
+      - audit_rules_suid_privilege_function.severity=high
+      - audit_sudo_log_events
+      - audit_sudo_log_events.severity=high
+      - audit_rules_time_adjtimex
+      - audit_rules_time_adjtimex.severity=high
+      - audit_rules_time_settimeofday
+      - audit_rules_time_settimeofday.severity=high
+      - audit_rules_time_clock_settime
+      - audit_rules_time_clock_settime.severity=high
+      - audit_rules_networkconfig_modification
+      - audit_rules_networkconfig_modification.severity=high
+      - audit_rules_privileged_commands
+      - audit_rules_privileged_commands.severity=high
+      - audit_rules_usergroup_modification_group
+      - audit_rules_usergroup_modification_group.severity=high
+      - audit_rules_usergroup_modification_passwd
+      - audit_rules_usergroup_modification_passwd.severity=high
+      - audit_rules_usergroup_modification_shadow
+      - audit_rules_usergroup_modification_shadow.severity=high
+      - audit_rules_dac_modification_chmod
+      - audit_rules_dac_modification_chmod.severity=high
+      - audit_rules_dac_modification_chown
+      - audit_rules_dac_modification_chown.severity=high
+      - audit_rules_file_deletion_events_unlink
+      - audit_rules_file_deletion_events_unlink.severity=high
+      - audit_rules_file_deletion_events_rename
+      - audit_rules_file_deletion_events_rename.severity=high
+      - audit_rules_mac_modification
+      - audit_rules_mac_modification.severity=high
+      - audit_rules_execution_chcon
+      - audit_rules_execution_chcon.severity=high
+      - audit_rules_kernel_module_loading_init
+      - audit_rules_kernel_module_loading_init.severity=high
+
+  - id: 2.3_journald_configured
+    title: Ensure Journald is Configured
+    levels:
+      - base
+    status: automated
+    rules:
+      - service_systemd-journald_enabled
+      - service_systemd-journald_enabled.severity=high
+      - journald_compress
+      - journald_compress.severity=low
+      - journald_storage
+      - journald_storage.severity=low
+
+  - id: 2.4_memory_protection_enabled
+    title: Ensure Memory Protection is Enabled
+    levels:
+      - base
+    status: automated
+    rules:
+      - sysctl_kernel_randomize_va_space
+      - sysctl_kernel_randomize_va_space.severity=high
+      - sysctl_kernel_exec_shield
+      - sysctl_kernel_exec_shield.severity=high
+      - sysctl_kernel_kptr_restrict
+      - sysctl_kernel_kptr_restrict.severity=high
+
+  - id: 2.5_cron_permissions_configured
+    title: Ensure Cron Permissions are Configured
+    levels:
+      - base
+    status: automated
+    rules:
+      - file_owner_crontab
+      - file_owner_crontab.severity=high
+      - file_groupowner_crontab
+      - file_groupowner_crontab.severity=high
+      - file_permissions_crontab
+      - file_permissions_crontab.severity=high
+      - file_owner_cron_allow
+      - file_owner_cron_allow.severity=high
+      - file_groupowner_cron_allow
+      - file_groupowner_cron_allow.severity=high
+      - file_permissions_cron_allow
+      - file_permissions_cron_allow.severity=high
+
+  - id: 2.6_ensure_a_single_time_synchronization_daemon_used
+    title: Ensure a single time synchronization daemon is in use (Automated)
+    levels:
+      - base
+    status: automated
+    rules:
+      - var_timesync_service=systemd-timesyncd
+      - package_timesyncd_installed
+      - service_timesyncd_enabled
+      - service_chronyd_disabled
+    notes: |
+      To select which timesync daemon to install and configure, use the
+      profile variable var_timesync_service.
+
+  - id: 2.7_service_minimization
+    title: Ensure Unnecessary Services are Removed
+    levels:
+      - base
+    status: automated
+    rules:
+      - package_xinetd_removed
+      - package_xinetd_removed.severity=low
+      - package_dhcp_removed
+      - package_dhcp_removed.severity=low
+      - package_bind_removed
+      - package_bind_removed.severity=low
+      - package_vsftpd_removed
+      - package_vsftpd_removed.severity=low
+      - package_tftp-server_removed
+      - package_tftp-server_removed.severity=low
+      - package_httpd_removed
+      - package_httpd_removed.severity=low
+      - package_nginx_removed
+      - package_nginx_removed.severity=low
+      - package_cyrus-imapd_removed
+      - package_cyrus-imapd_removed.severity=low
+      - package_dovecot_removed
+      - package_dovecot_removed.severity=low
+      - package_samba_removed
+      - package_samba_removed.severity=low
+      - package_squid_removed
+      - package_squid_removed.severity=low
+      - package_net-snmp_removed
+      - package_net-snmp_removed.severity=low
+      - package_ypserv_removed
+      - package_ypserv_removed.severity=low
+      - package_telnet_removed
+      - package_telnet_removed.severity=low
+      - package_telnet-server_removed
+      - package_telnet-server_removed.severity=low
+      - package_rsync_removed
+      - package_rsync_removed.severity=low
+      - package_rsh_removed
+      - package_rsh_removed.severity=low
+      - package_rsh-server_removed
+      - package_rsh-server_removed.severity=low
+      - package_sendmail_removed
+      - package_sendmail_removed.severity=low
+      - package_ypbind_removed
+      - package_ypbind_removed.severity=low
+      - package_talk-server_removed
+      - package_talk-server_removed.severity=low
+      - package_talk_removed
+      - package_talk_removed.severity=low
+      - service_nfs_disabled
+      - service_nfs_disabled.severity=low
+      - service_rpcbind_disabled
+      - service_rpcbind_disabled.severity=low
+
+  - id: 3.1_sshd_hardening
+    title: Ensure SSHD is Hardened
+    levels:
+      - base
+    status: automated
+    rules:
+      - file_owner_sshd_config
+      - file_owner_sshd_config.severity=high
+      - file_groupowner_sshd_config
+      - file_groupowner_sshd_config.severity=high
+      - file_permissions_sshd_config
+      - file_permissions_sshd_config.severity=high
+      - sshd_set_loglevel_verbose
+      - sshd_set_loglevel_verbose.severity=high
+      - sshd_enable_pam
+      - sshd_enable_pam.severity=high
+      - sshd_disable_root_login
+      - sshd_disable_root_login.severity=high
+      - disable_host_auth
+      - disable_host_auth.severity=high
+      - sshd_disable_empty_passwords
+      - sshd_disable_empty_passwords.severity=high
+      - sshd_do_not_permit_user_env
+      - sshd_do_not_permit_user_env.severity=high
+      - sshd_disable_rhosts
+      - sshd_disable_rhosts.severity=high
+      - sshd_disable_x11_forwarding
+      - sshd_disable_x11_forwarding.severity=high
+      - sshd_disable_tcp_forwarding
+      - sshd_disable_tcp_forwarding.severity=high
+      - sshd_set_max_auth_tries
+      - sshd_set_max_auth_tries.severity=high
+      - var_sshd_set_maxstartups=10:30:60
+      - var_sshd_max_sessions=10
+      - var_sshd_set_login_grace_time=60
+      - sshd_set_idle_timeout
+      - sshd_set_idle_timeout.severity=high
+      - sshd_set_keepalive
+      - sshd_set_keepalive.severity=high
+      - var_rekey_limit_size=1G
+      - var_rekey_limit_time=1hour
+      - sshd_use_strong_rng
+      - sshd_use_strong_rng.severity=high
+
+  - id: 4.1_network_hardening
+    title: Ensure Network is Hardened
+    levels:
+      - base
+    status: automated
+    rules:
+      - kernel_module_sctp_disabled
+      - kernel_module_sctp_disabled.severity=low
+      - kernel_module_dccp_disabled
+      - kernel_module_dccp_disabled.severity=low
+      - sysctl_net_ipv4_conf_all_send_redirects
+      - sysctl_net_ipv4_conf_all_send_redirects.severity=high
+      - sysctl_net_ipv4_conf_default_send_redirects
+      - sysctl_net_ipv4_conf_default_send_redirects.severity=high
+      - sysctl_net_ipv4_conf_all_accept_source_route
+      - sysctl_net_ipv4_conf_all_accept_source_route.severity=high
+      - sysctl_net_ipv4_conf_default_accept_source_route
+      - sysctl_net_ipv4_conf_default_accept_source_route.severity=high
+      - sysctl_net_ipv6_conf_all_accept_source_route
+      - sysctl_net_ipv6_conf_all_accept_source_route.severity=high
+      - sysctl_net_ipv6_conf_default_accept_source_route
+      - sysctl_net_ipv6_conf_default_accept_source_route.severity=high
+      - sysctl_net_ipv4_conf_all_accept_redirects
+      - sysctl_net_ipv4_conf_all_accept_redirects.severity=high
+      - sysctl_net_ipv4_conf_default_accept_redirects
+      - sysctl_net_ipv4_conf_default_accept_redirects.severity=high
+      - sysctl_net_ipv6_conf_all_accept_redirects
+      - sysctl_net_ipv6_conf_all_accept_redirects.severity=high
+      - sysctl_net_ipv6_conf_default_accept_redirects
+      - sysctl_net_ipv6_conf_default_accept_redirects.severity=high
+      - sysctl_net_ipv4_conf_all_secure_redirects
+      - sysctl_net_ipv4_conf_all_secure_redirects.severity=high
+      - sysctl_net_ipv4_conf_default_secure_redirects
+      - sysctl_net_ipv4_conf_default_secure_redirects.severity=high
+      - sysctl_net_ipv4_conf_all_log_martians
+      - sysctl_net_ipv4_conf_all_log_martians.severity=high
+      - sysctl_net_ipv4_conf_default_log_martians
+      - sysctl_net_ipv4_conf_default_log_martians.severity=high
+      - sysctl_net_ipv4_icmp_echo_ignore_broadcasts
+      - sysctl_net_ipv4_icmp_echo_ignore_broadcasts.severity=high
+      - sysctl_net_ipv4_icmp_ignore_bogus_error_responses
+      - sysctl_net_ipv4_icmp_ignore_bogus_error_responses.severity=high
+      - sysctl_net_ipv4_conf_all_rp_filter
+      - sysctl_net_ipv4_conf_all_rp_filter.severity=high
+      - sysctl_net_ipv4_conf_default_rp_filter
+      - sysctl_net_ipv4_conf_default_rp_filter.severity=high
+      - sysctl_net_ipv4_tcp_syncookies
+      - sysctl_net_ipv4_tcp_syncookies.severity=high
+
+  - id: 4.2_firewalld_configured
+    title: Ensure Firewalld is Configured
+    levels:
+      - base
+    status: automated
+    rules:
+      - package_firewalld_installed
+      - package_firewalld_installed.severity=high
+      - service_firewalld_enabled
+      - service_firewalld_enabled.severity=high
+      - service_nftables_disabled
+      - service_nftables_disabled.severity=high
+
+  - id: 5.1_selinux_enforcing
+    title: Ensure SELinux is Enforcing
+    levels:
+      - base
+    status: automated
+    rules:
+      - package_libselinux_installed
+      - package_libselinux_installed.severity=high
+      - grub2_enable_selinux
+      - grub2_enable_selinux.severity=high
+      - selinux_policytype
+      - selinux_policytype.severity=high
+      - var_selinux_policy_name=targeted
+      - selinux_state
+      - selinux_state.severity=high
+      - package_mcstrans_removed
+      - package_mcstrans_removed.severity=high
+      - sysctl_fs_protected_hardlinks
+      - sysctl_fs_protected_hardlinks.severity=high
+      - sysctl_fs_protected_symlinks
+      - sysctl_fs_protected_symlinks.severity=high
+
+  - id: 5.2_periodic_compliance_scans
+    title: Ensure Periodic Compliance Scans are Performed
+    levels:
+      - base
+    status: manual
diff --git a/linux_os/guide/services/obsolete/service_rsyncd_disabled/rule.yml b/linux_os/guide/services/obsolete/service_rsyncd_disabled/rule.yml
index 1db77bb04a..d2bc832377 100644
--- a/linux_os/guide/services/obsolete/service_rsyncd_disabled/rule.yml
+++ b/linux_os/guide/services/obsolete/service_rsyncd_disabled/rule.yml
@@ -43,5 +43,6 @@ template:
         packagename@sle15: rsync
         packagename@slmicro5: rsync
         packagename@openeuler2203: rsync
+        packagename@openruyi: rsync
         servicename@ubuntu2404: rsync
         packagename@ubuntu2404: rsync
diff --git a/linux_os/guide/services/ssh/sshd_strong_kex.var b/linux_os/guide/services/ssh/sshd_strong_kex.var
index d4a78a3fe4..373ef2ff49 100644
--- a/linux_os/guide/services/ssh/sshd_strong_kex.var
+++ b/linux_os/guide/services/ssh/sshd_strong_kex.var
@@ -21,4 +21,5 @@ options:
     cis_ubuntu2204: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256
     cis_ubuntu2404: sntrup761x25519-sha512@openssh.com,curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256
     std_openeuler: curve25519-sha256,curve25519-sha256@libssh.org,diffie-hellman-group-exchange-sha256
+    std_openruyi: curve25519-sha256,curve25519-sha256@libssh.org,diffie-hellman-group-exchange-sha256
     cis_debian12: sntrup761x25519-sha512@openssh.com,curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256
diff --git a/linux_os/guide/system/accounts/accounts-physical/require_emergency_target_auth/oval/shared.xml b/linux_os/guide/system/accounts/accounts-physical/require_emergency_target_auth/oval/shared.xml
index b359d976e0..ec09a7b47a 100644
--- a/linux_os/guide/system/accounts/accounts-physical/require_emergency_target_auth/oval/shared.xml
+++ b/linux_os/guide/system/accounts/accounts-physical/require_emergency_target_auth/oval/shared.xml
@@ -35,7 +35,7 @@
   </ind:textfilecontent54_test>
   <ind:textfilecontent54_object id="obj_require_emergency_service" version="1">
     <ind:filepath>/usr/lib/systemd/system/emergency.service</ind:filepath>
-    {{%- if product in ["fedora", "ol8", "ol9", "kylinserver10", "openeuler2203", "sle12", "sle15", "sle16", "slmicro5"] or 'rhel' in product -%}}
+    {{%- if product in ["fedora", "ol8", "ol9", "kylinserver10", "openeuler2203", "openruyi", "sle12", "sle15", "sle16", "slmicro5"] or 'rhel' in product -%}}
     <ind:pattern operation="pattern match">^ExecStart=\-/usr/lib/systemd/systemd-sulogin-shell[\s]+emergency</ind:pattern>
     {{%- else -%}}
     <ind:pattern operation="pattern match">^ExecStart=\-/bin/sh[\s]+-c[\s]+\"(/usr)?/sbin/sulogin;[\s]+/usr/bin/systemctl[\s]+--fail[\s]+--no-block[\s]+default\"</ind:pattern>
diff --git a/linux_os/guide/system/accounts/accounts-physical/require_emergency_target_auth/rule.yml b/linux_os/guide/system/accounts/accounts-physical/require_emergency_target_auth/rule.yml
index 00553417e3..1ecbf61cc2 100644
--- a/linux_os/guide/system/accounts/accounts-physical/require_emergency_target_auth/rule.yml
+++ b/linux_os/guide/system/accounts/accounts-physical/require_emergency_target_auth/rule.yml
@@ -77,7 +77,7 @@ fixtext: |-
     Configure {{{ full_name }}} to require authentication for system emergency mode.
 
     Add or edit the following line in "/usr/lib/systemd/system/emergency.service":
-    {{% if product in ["fedora", "ol8", "ol9", "kylinserver10", "openeuler2203", "rhel8", "rhel9", "sle12", "sle15", "sle16", "slmicro5"] -%}}
+    {{% if product in ["fedora", "ol8", "ol9", "kylinserver10", "openeuler2203", "openruyi", "rhel8", "rhel9", "sle12", "sle15", "sle16", "slmicro5"] -%}}
     ExecStart=-/usr/lib/systemd/systemd-sulogin-shell emergency
     {{%- else -%}}
     ExecStart=-/bin/sh -c "/sbin/sulogin; /usr/bin/systemctl --fail --no-block default"
diff --git a/linux_os/guide/system/bootloader-grub2/non-uefi/grub2_password/tests/invalid_username.fail.sh b/linux_os/guide/system/bootloader-grub2/non-uefi/grub2_password/tests/invalid_username.fail.sh
index 663a3282cf..b6f05c2bc1 100644
--- a/linux_os/guide/system/bootloader-grub2/non-uefi/grub2_password/tests/invalid_username.fail.sh
+++ b/linux_os/guide/system/bootloader-grub2/non-uefi/grub2_password/tests/invalid_username.fail.sh
@@ -7,7 +7,7 @@
 
 make_grub_password
 
-{{% if 'ubuntu' in product %}}
+{{% if 'ubuntu' in product or product == 'openruyi' %}}
 test -n "$GRUB_CFG_ROOT" || GRUB_CFG_ROOT=/boot/grub
 {{% else %}}
 test -n "$GRUB_CFG_ROOT" || GRUB_CFG_ROOT=/boot/grub2
diff --git a/linux_os/guide/system/kernel_build_config/kernel_config_default_mmap_min_addr/oval/shared.xml b/linux_os/guide/system/kernel_build_config/kernel_config_default_mmap_min_addr/oval/shared.xml
index 7f2f865631..577cd8a84e 100644
--- a/linux_os/guide/system/kernel_build_config/kernel_config_default_mmap_min_addr/oval/shared.xml
+++ b/linux_os/guide/system/kernel_build_config/kernel_config_default_mmap_min_addr/oval/shared.xml
@@ -1,7 +1,7 @@
 <def-group>
   <definition class="compliance" id="kernel_config_default_mmap_min_addr"
   version="1">
-    {{{ oval_metadata("The kernel config CONFIG_DEFAULT_MMAP_MIN_ADDR should have value 65536 on x86_64 and 32768 on aarch64", affected_platforms=["multi_platform_all"], rule_title=rule_title) }}}
+    {{{ oval_metadata("The kernel config CONFIG_DEFAULT_MMAP_MIN_ADDR should have value 65536 on x86_64 and 32768 on aarch64 or riscv64", affected_platforms=["multi_platform_all"], rule_title=rule_title) }}}
     <criteria operator="AND">
       <criteria operator="OR">
         <criteria operator="AND">
@@ -16,6 +16,12 @@
           <criterion comment="Check presence of build configuration of installed kernels"
           test_ref="test_kernel_config_default_mmap_min_addr_aarch64" />
         </criteria>
+        <criteria operator="AND">
+          <criterion comment="Check architecture is riscv64"
+          test_ref="test_proc_sys_kernel_osrelease_arch_riscv64" />
+          <criterion comment="Check presence of build configuration of installed kernels"
+          test_ref="test_kernel_config_default_mmap_min_addr_riscv64" />
+        </criteria>
       </criteria>
       <criterion comment="Ensure all kernels have the config"
       test_ref="test_kernel_config_default_mmap_min_addr_all_kernels" />
@@ -34,6 +40,12 @@
     <ind:object object_ref="object_kernel_config_default_mmap_min_addr" />
     <ind:state state_ref="state_kernel_config_default_mmap_min_addr_aarch64" />
   </ind:textfilecontent54_test>
+  <ind:textfilecontent54_test check="all"
+  comment="Check /boot/config-.* files for CONFIG_DEFAULT_MMAP_MIN_ADDR=32768"
+  id="test_kernel_config_default_mmap_min_addr_riscv64" version="1">
+    <ind:object object_ref="object_kernel_config_default_mmap_min_addr" />
+    <ind:state state_ref="state_kernel_config_default_mmap_min_addr_riscv64" />
+  </ind:textfilecontent54_test>
 
   <ind:textfilecontent54_object id="object_kernel_config_default_mmap_min_addr" version="1">
     <ind:filepath operation="pattern match">^/boot/config-.*$</ind:filepath>
@@ -47,6 +59,9 @@
   <ind:textfilecontent54_state id="state_kernel_config_default_mmap_min_addr_aarch64" version="1">
     <ind:subexpression operation="equals" datatype="string">32768</ind:subexpression>
   </ind:textfilecontent54_state>
+  <ind:textfilecontent54_state id="state_kernel_config_default_mmap_min_addr_riscv64" version="1">
+    <ind:subexpression operation="equals" datatype="string">32768</ind:subexpression>
+  </ind:textfilecontent54_state>
 
   <ind:variable_test check="all" check_existence="all_exist"
   comment="Check if all installed kernels are compliant"
@@ -89,6 +104,23 @@
       <unix:filename operation="pattern match">^config-.*$</unix:filename>
   </unix:file_object>
 
+  <ind:textfilecontent54_test check="all" check_existence="all_exist"
+      comment="proc_sys_kernel is for riscv64 architecture"
+      id="test_proc_sys_kernel_osrelease_arch_riscv64"
+  version="1">
+    <ind:object object_ref="object_proc_sys_kernel_osrelease_arch_riscv64" />
+    <ind:state state_ref="state_proc_sys_kernel_osrelease_arch_riscv64" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="object_proc_sys_kernel_osrelease_arch_riscv64" version="1">
+    <ind:filepath>/proc/sys/kernel/osrelease</ind:filepath>
+    <ind:pattern operation="pattern match">^.*\.(.*)$</ind:pattern>
+    <ind:instance datatype="int" operation="greater than or equal">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <ind:textfilecontent54_state id="state_proc_sys_kernel_osrelease_arch_riscv64" version="1">
+    <ind:subexpression datatype="string" operation="pattern match">^riscv64$</ind:subexpression>
+  </ind:textfilecontent54_state>
 
   <ind:textfilecontent54_test check="all" check_existence="all_exist"
       comment="proc_sys_kernel is for aarch64 architecture"
diff --git a/linux_os/guide/system/permissions/files/file_empty_link_prohibit/rule.yml b/linux_os/guide/system/permissions/files/file_empty_link_prohibit/rule.yml
new file mode 100644
index 0000000000..b1db0e6cc0
--- /dev/null
+++ b/linux_os/guide/system/permissions/files/file_empty_link_prohibit/rule.yml
@@ -0,0 +1,26 @@
+documentation_complete: true
+
+
+title: 'Empty link files are prohibited'
+
+description: |-
+    <p><tt>It can not be scanned automatically, please check it manually.</tt></p>
+    <p>If any symlink files have no camonical path, it should be removed.</p>
+    <ul>
+    <li>To find broken symlinks in the current directory:
+    <pre># find ./ -xtype l</pre>
+    </li>
+    <li>To find system-wide broken symlinks (excluding virtual filesystems):
+    <pre># find / \( -path /proc -o -path /sys -o -path /dev -o -path /run -o -path /var \) -prune -o -xtype l -print</pre>
+    </li>
+    <li>To find broken symlinks only on the same disk partition:
+    <pre># find / -xdev -xtype l</pre>
+    </li>
+    </ul>
+
+rationale: |-
+    Broken symlinks clutter the system, complicate maintenance, and can cause application
+    errors or unexpected behavior. Removing them ensures a cleaner, more predictable, and
+    maintainable environment.
+
+severity: high
diff --git a/linux_os/guide/system/permissions/files/file_empty_link_prohibit/sce/shared.sh b/linux_os/guide/system/permissions/files/file_empty_link_prohibit/sce/shared.sh
new file mode 100644
index 0000000000..05a047f37b
--- /dev/null
+++ b/linux_os/guide/system/permissions/files/file_empty_link_prohibit/sce/shared.sh
@@ -0,0 +1,11 @@
+#!/bin/bash
+# platform = multi_platform_openruyi
+# check-import = stdout
+
+temp=$(find / \( -path /proc -o -path /sys -o -path /dev -o -path /run -o -path /var \) -prune -o -xtype l -print 2>/dev/null)
+
+if [ -z "$temp" ]; then
+    exit "$XCCDF_RESULT_PASS"
+fi
+
+exit "$XCCDF_RESULT_FAIL"
diff --git a/linux_os/guide/system/permissions/files/file_hidden_executable_prohibit/rule.yml b/linux_os/guide/system/permissions/files/file_hidden_executable_prohibit/rule.yml
new file mode 100644
index 0000000000..e43a9ba3ae
--- /dev/null
+++ b/linux_os/guide/system/permissions/files/file_hidden_executable_prohibit/rule.yml
@@ -0,0 +1,25 @@
+documentation_complete: true
+
+
+title: 'Disallow hidden executable files'
+
+description: |-
+    In the Linux system, files prefixed with "." are hidden files (except for "." and ".." in the
+    current directory and upper directory), and executable hidden files are not allowed in the system.
+
+    The three files .bashrc, .bash_profile, and .bash_logout are the script files used when the system
+    logs in/out of the shell after creating a user account. They are in line with industry practice
+    and do not need to be deleted. Other hidden executable files must have their execute permissions
+    removed or be deleted.
+
+    To find hidden files with execute permissions:
+    <pre># find / \( -path /proc -o -path /sys -o -path /dev -o -path /run -o -path /var \) -prune -o -type f -name ".*" -perm /111 ! -name ".bashrc" ! -name ".bash_profile" ! -name ".bash_logout" -print</pre>
+
+rationale: |-
+     Hidden files (prefixed with a dot) are not visible in standard directory listings, making them
+     ideal for hiding malware or backdoors. Allowing execute permissions on them increases the risk
+     of undetected malicious code execution. Prohibiting executable hidden files (with standard
+     exceptions like shell configs) reduces the attack surface and enforces least privilege.
+
+severity: high
+
diff --git a/linux_os/guide/system/permissions/files/file_hidden_executable_prohibit/sce/shared.sh b/linux_os/guide/system/permissions/files/file_hidden_executable_prohibit/sce/shared.sh
new file mode 100644
index 0000000000..26a0969e09
--- /dev/null
+++ b/linux_os/guide/system/permissions/files/file_hidden_executable_prohibit/sce/shared.sh
@@ -0,0 +1,13 @@
+#!/bin/bash
+# platform = multi_platform_openruyi
+# check-import = stdout
+
+exit "$XCCDF_RESULT_FAIL"
+
+temp=$(find / \( -path /proc -o -path /sys -o -path /dev -o -path /run -o -path /var \) -prune -o -type f -name ".*" -perm /111 ! -name ".bashrc" ! -name ".bash_profile" ! -name ".bash_logout" -print 2>/dev/null)
+
+if [ -z "$temp" ]; then
+    exit "$XCCDF_RESULT_PASS"
+fi
+
+exit "$XCCDF_RESULT_FAIL"
diff --git a/product_properties/10-grub.yml b/product_properties/10-grub.yml
index b2c17c2335..f406274e58 100644
--- a/product_properties/10-grub.yml
+++ b/product_properties/10-grub.yml
@@ -29,3 +29,8 @@ overrides:
   grub2_uefi_boot_path: "/boot/grub"
 {{% endif %}}
 {{% endif %}}
+{{% if product == "openruyi" %}}
+  grub2_boot_path: "/boot/grub"
+  grub2_uefi_boot_path: "/boot/grub"
+  grub_helper_executable: "grub-mkconfig"
+{{% endif %}}
diff --git a/products/openruyi/CMakeLists.txt b/products/openruyi/CMakeLists.txt
new file mode 100644
index 0000000000..bb9c6b7b62
--- /dev/null
+++ b/products/openruyi/CMakeLists.txt
@@ -0,0 +1,6 @@
+# Sometimes our users will try to do: "cd openruyi; cmake ." That needs to error in a nice way.
+if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
+    message(FATAL_ERROR "cmake has to be used on the root CMakeLists.txt, see the Building ComplianceAsCode section in the Developer Guide!")
+endif()
+
+ssg_build_product("openruyi")
diff --git a/products/openruyi/product.yml b/products/openruyi/product.yml
new file mode 100644
index 0000000000..c9791cfa55
--- /dev/null
+++ b/products/openruyi/product.yml
@@ -0,0 +1,24 @@
+product: openruyi
+full_name: openRuyi
+arch: riscv64
+type: platform
+
+benchmark_id: OPENRUYI
+benchmark_root: "../../linux_os/guide"
+
+profiles_root: "./profiles"
+
+pkg_manager: "dnf"
+pkg_manager_config_file: "/etc/yum.conf"
+
+init_system: "systemd"
+
+cpes_root: "../../shared/applicability"
+cpes:
+  - openruyi:
+      name: "cpe:/o:openruyi:openruyi:creek"
+      title: "openRuyi Creek"
+      check_id: installed_OS_is_openruyi_riscv
+
+reference_uris:
+  cis: 'https://workbench.cisecurity.org/communities/101'
\ No newline at end of file
diff --git a/products/openruyi/profiles/standard.profile b/products/openruyi/profiles/standard.profile
new file mode 100644
index 0000000000..740ab7c66a
--- /dev/null
+++ b/products/openruyi/profiles/standard.profile
@@ -0,0 +1,14 @@
+documentation_complete: true
+
+metadata:
+    version: 1.0
+
+title: 'Standard System Security Profile for openRuyi'
+
+description: |-
+    This profile contains rules to ensure standard security baseline
+    of an openRuyi system. Regardless of your system's workload
+    all of these checks should pass.
+
+selections:
+    - std_openruyi:all:base
\ No newline at end of file
diff --git a/shared/applicability/bootc.yml b/shared/applicability/bootc.yml
index 1b7cabf533..38969bc682 100644
--- a/shared/applicability/bootc.yml
+++ b/shared/applicability/bootc.yml
@@ -17,6 +17,8 @@ title: Bootable containers
 check_id: bootc
 {{% if product == "fedora" or "rhel" in product %}}
 {{% set kernel_package = "kernel-core" %}}
+{{% elif product == "openruyi" %}}
+{{% set kernel_package = "linux-core" %}}
 {{% else %}}
 {{% set kernel_package = "kernel" %}}
 {{% endif %}}
diff --git a/shared/applicability/oval/bootc.xml b/shared/applicability/oval/bootc.xml
index 4eaf3009bf..4de544308c 100644
--- a/shared/applicability/oval/bootc.xml
+++ b/shared/applicability/oval/bootc.xml
@@ -14,6 +14,8 @@
   </definition>
 {{% if product == "fedora" or "rhel" in product %}}
 {{% set kernel_package = "kernel-core" %}}
+{{% elif product == "openruyi" %}}
+{{% set kernel_package = "linux-core" %}}
 {{% else %}}
 {{% set kernel_package = "kernel" %}}
 {{% endif %}}
diff --git a/shared/applicability/oval/installed_OS_is_openruyi_riscv.xml b/shared/applicability/oval/installed_OS_is_openruyi_riscv.xml
new file mode 100644
index 0000000000..4383ce6d59
--- /dev/null
+++ b/shared/applicability/oval/installed_OS_is_openruyi_riscv.xml
@@ -0,0 +1,25 @@
+<def-group>
+  <definition class="inventory" id="installed_OS_is_openruyi_riscv" version="3">
+    <metadata>
+      <title>Installed operating system is openRuyi (RISCV64)</title>
+      <affected family="unix">
+        <platform>multi_platform_all</platform>
+      </affected>
+      <reference ref_id="cpe:/o:openruyi:openruyi:creek" source="CPE" />
+      <description>The operating system installed on the system is openRuyi (RISCV64)</description>
+    </metadata>
+    <criteria operator="AND">
+      <extend_definition comment="Installed OS is part of the Unix family" definition_ref="installed_OS_is_part_of_Unix_family" />
+      <criterion comment="openruyi-release RPM packages are installed" test_ref="test_openruyi_release_rpm" />
+      <extend_definition comment="Architecture is riscv64" definition_ref="proc_sys_kernel_osrelease_arch_riscv64" />
+    </criteria>
+  </definition>
+
+  <linux:rpminfo_test check="all" check_existence="at_least_one_exists" comment="openruyi-release RPM packages are installed" id="test_openruyi_release_rpm" version="1">
+    <linux:object object_ref="object_openruyi_release_rpm" />
+  </linux:rpminfo_test>
+  <linux:rpminfo_object id="object_openruyi_release_rpm" version="1">
+    <linux:name operation="pattern match">open[rR]uyi-release.*</linux:name>
+  </linux:rpminfo_object>
+
+</def-group>
diff --git a/shared/applicability/oval/installed_env_has_grub2_package.xml b/shared/applicability/oval/installed_env_has_grub2_package.xml
index 827ea9f193..1845b99354 100644
--- a/shared/applicability/oval/installed_env_has_grub2_package.xml
+++ b/shared/applicability/oval/installed_env_has_grub2_package.xml
@@ -11,6 +11,7 @@
     </metadata>
     <criteria operator="AND">
       <criterion comment="Package grub2-common is installed" test_ref="test_env_has_grub2_installed" />
+      {{% if product != "openruyi" %}}
       <criteria operator="OR">
         <!-- On ppc64le systems, OF (Grub2) or OPAL (petitboot) can be used,
         and unfortunately the package set installed is the same in both cases.
@@ -21,6 +22,7 @@
         <!-- or when grub2-common is installed, arch is a ppc64le but OPAL is not used -->
         <criterion comment="Test if OPAL is not used" test_ref="test_system_using_opal" negate="true" />
       </criteria>
+      {{% endif %}}
     </criteria>
   </definition>
 
@@ -31,7 +33,7 @@
     <linux:object object_ref="obj_env_has_grub2_installed" />
   </linux:rpminfo_test>
   <linux:rpminfo_object id="obj_env_has_grub2_installed" version="1">
-    <linux:name>{{% if product in ["sle12", "sle15", "slmicro5", "slmicro6"] %}}grub2{{% else %}}grub2-common{{% endif %}}</linux:name>
+    <linux:name>{{% if product in ["sle12", "sle15", "slmicro5", "slmicro6"] %}}grub2{{% elif product == "openruyi" %}}grub{{% else %}}grub2-common{{% endif %}}</linux:name>
   </linux:rpminfo_object>
 {{% elif pkg_system == "dpkg" %}}
   <linux:dpkginfo_test check="all" check_existence="all_exist"
diff --git a/shared/applicability/oval/proc_sys_kernel_osrelease_arch_riscv64.xml b/shared/applicability/oval/proc_sys_kernel_osrelease_arch_riscv64.xml
new file mode 100644
index 0000000000..128ca33f2c
--- /dev/null
+++ b/shared/applicability/oval/proc_sys_kernel_osrelease_arch_riscv64.xml
@@ -0,0 +1 @@
+{{{ oval_check_proc_sys_kernel_osrelease_arch("riscv64") }}}
diff --git a/shared/applicability/oval/system_with_kernel.xml b/shared/applicability/oval/system_with_kernel.xml
index d1cbf79c4d..202ff63a79 100644
--- a/shared/applicability/oval/system_with_kernel.xml
+++ b/shared/applicability/oval/system_with_kernel.xml
@@ -16,6 +16,8 @@
 {{% elif 'sle' in product or 'slmicro' in product %}}
 {{{ oval_test_package_installed(package="kernel-default", test_id="inventory_test_kernel_installed") }}}
 {{{ oval_test_package_installed(package="kernel-default-base", test_id="inventory_test_kernel_default_base_installed") }}}
+{{% elif product == 'openruyi' %}}
+{{{ oval_test_package_installed(package="linux-core", test_id="inventory_test_kernel_installed") }}}
 {{% else %}}
 {{{ oval_test_package_installed(package="kernel-core", test_id="inventory_test_kernel_installed") }}}
 {{% endif %}}
diff --git a/shared/applicability/package.yml b/shared/applicability/package.yml
index 75cd60e08c..52e5d19f12 100644
--- a/shared/applicability/package.yml
+++ b/shared/applicability/package.yml
@@ -35,7 +35,7 @@ args:
   chrony:
     pkgname: chrony
   dnf:
-    {{% if product == "fedora" %}}
+    {{% if product in ["fedora", "openruyi"] %}}
     pkgname: dnf5
     {{% else %}}
     pkgname: dnf
@@ -113,7 +113,7 @@ args:
     pkgname: rsh-server
   shadow-utils:
     {{% if pkg_system == "rpm" %}}
-    {{% if 'sle' in product or 'slmicro' in product or product in ["kylinserver10", "openeuler2203"] %}}
+    {{% if 'sle' in product or 'slmicro' in product or product in ["kylinserver10", "openeuler2203", "openruyi"] %}}
       pkgname: shadow
     {{% else %}}
       pkgname: shadow-utils
diff --git a/shared/applicability/riscv64_arch.yml b/shared/applicability/riscv64_arch.yml
new file mode 100644
index 0000000000..a1da619758
--- /dev/null
+++ b/shared/applicability/riscv64_arch.yml
@@ -0,0 +1,5 @@
+name: cpe:/a:riscv64_arch
+title: System architecture is RISCV64
+check_id: proc_sys_kernel_osrelease_arch_riscv64
+bash_conditional: {{{ bash_arch_conditional("riscv64") }}}
+ansible_conditional: 'ansible_architecture == "riscv64"'
diff --git a/shared/applicability/system_with_kernel.yml b/shared/applicability/system_with_kernel.yml
index 07793b1a1a..98c8f50627 100644
--- a/shared/applicability/system_with_kernel.yml
+++ b/shared/applicability/system_with_kernel.yml
@@ -22,6 +22,8 @@ bash_conditional: "rpm --quiet -q kernel-default || rpm --quiet -q kernel-defaul
 bash_conditional: "rpm --quiet -q kernel || rpm --quiet -q kernel-uek"
 {{% elif product == "fedora" or "rhel" in product %}}
 bash_conditional: "rpm --quiet -q kernel-core"
+{{% elif product == "openruyi" %}}
+bash_conditional: "rpm --quiet -q linux-core"
 {{% else %}}
 bash_conditional: "rpm --quiet -q kernel"
 {{% endif %}}
@@ -40,6 +42,8 @@ ansible_conditional: '("kernel-default" in ansible_facts.packages or "kernel-def
 ansible_conditional: '("kernel" in ansible_facts.packages or "kernel-uek" in ansible_facts.packages)'
 {{% elif product == "fedora" or "rhel" in product %}}
 ansible_conditional: '"kernel-core" in ansible_facts.packages'
+{{% elif product == "openruyi" %}}
+ansible_conditional: '"linux-core" in ansible_facts.packages'
 {{% else %}}
 ansible_conditional: '"kernel" in ansible_facts.packages'
 {{% endif %}}
diff --git a/shared/checks/oval/bootc.xml b/shared/checks/oval/bootc.xml
index 4eaf3009bf..4de544308c 100644
--- a/shared/checks/oval/bootc.xml
+++ b/shared/checks/oval/bootc.xml
@@ -14,6 +14,8 @@
   </definition>
 {{% if product == "fedora" or "rhel" in product %}}
 {{% set kernel_package = "kernel-core" %}}
+{{% elif product == "openruyi" %}}
+{{% set kernel_package = "linux-core" %}}
 {{% else %}}
 {{% set kernel_package = "kernel" %}}
 {{% endif %}}
diff --git a/shared/checks/oval/installed_OS_is_openruyi.xml b/shared/checks/oval/installed_OS_is_openruyi.xml
new file mode 100644
index 0000000000..dd87dddf9c
--- /dev/null
+++ b/shared/checks/oval/installed_OS_is_openruyi.xml
@@ -0,0 +1,24 @@
+<def-group>
+  <definition class="inventory" id="installed_OS_is_openruyi" version="3">
+    <metadata>
+      <title>Installed operating system is openRuyi</title>
+      <affected family="unix">
+        <platform>multi_platform_all</platform>
+      </affected>
+      <reference ref_id="cpe:/o:openruyi:openruyi:creek" source="CPE" />
+      <description>The operating system installed on the system is openRuyi</description>
+    </metadata>
+    <criteria operator="AND">
+      <extend_definition comment="Installed OS is part of the Unix family" definition_ref="installed_OS_is_part_of_Unix_family" />
+      <criterion comment="openruyi-release RPM packages are installed" test_ref="test_openruyi_release_rpm" />
+    </criteria>
+  </definition>
+
+  <linux:rpminfo_test check="all" check_existence="at_least_one_exists" comment="openruyi-release RPM packages are installed" id="test_openruyi_release_rpm" version="1">
+    <linux:object object_ref="object_openruyi_release_rpm" />
+  </linux:rpminfo_test>
+  <linux:rpminfo_object id="object_openruyi_release_rpm" version="1">
+    <linux:name operation="pattern match">open[rR]uyi-release.*</linux:name>
+  </linux:rpminfo_object>
+
+</def-group>
diff --git a/shared/checks/oval/sysctl_kernel_ipv6_disable.xml b/shared/checks/oval/sysctl_kernel_ipv6_disable.xml
index b718ded260..9cf1aee4e7 100644
--- a/shared/checks/oval/sysctl_kernel_ipv6_disable.xml
+++ b/shared/checks/oval/sysctl_kernel_ipv6_disable.xml
@@ -12,6 +12,7 @@
 	<platform>multi_platform_kylinserver</platform>
 	<platform>multi_platform_openembedded</platform>
 	<platform>multi_platform_openeuler</platform>
+	<platform>multi_platform_openruyi</platform>
 	<platform>multi_platform_opensuse</platform>
 	<platform>multi_platform_ol</platform>
 	<platform>multi_platform_rhcos</platform>
diff --git a/shared/checks/oval/system_info_architecture_64bit.xml b/shared/checks/oval/system_info_architecture_64bit.xml
index b6dac63f0d..c46797f414 100644
--- a/shared/checks/oval/system_info_architecture_64bit.xml
+++ b/shared/checks/oval/system_info_architecture_64bit.xml
@@ -19,6 +19,8 @@
       definition_ref="system_info_architecture_aarch_64" />
       <extend_definition comment="Generic test for s390x architecture"
       definition_ref="system_info_architecture_s390_64" />
+      <extend_definition comment="Generic test for riscv64 architecture"
+      definition_ref="system_info_architecture_riscv_64" />
     </criteria>
   </definition>
 
diff --git a/shared/checks/oval/system_info_architecture_riscv_64.xml b/shared/checks/oval/system_info_architecture_riscv_64.xml
new file mode 100644
index 0000000000..26e2498a3e
--- /dev/null
+++ b/shared/checks/oval/system_info_architecture_riscv_64.xml
@@ -0,0 +1,29 @@
+<def-group>
+  <definition class="compliance" id="system_info_architecture_riscv_64"
+  version="1">
+    <!-- Note that this does not meet requirements for class=inventory as
+         that only tests for patches per 5.10.1 Revision 1 -->
+    <metadata>
+      <title>Test for riscv_64 Architecture</title>
+      <affected family="unix">
+        <platform>multi_platform_all</platform>
+      </affected>
+      <description>Generic test for riscv_64 architecture to be used by other tests</description>
+    </metadata>
+    <criteria>
+      <criterion comment="Generic test for riscv_64 architecture"
+      test_ref="test_system_info_architecture_riscv_64" />
+    </criteria>
+  </definition>
+  <unix:uname_test check="all" comment="64 bit architecture"
+  id="test_system_info_architecture_riscv_64" version="1">
+    <unix:object object_ref="object_system_info_architecture_riscv_64" />
+    <unix:state state_ref="state_system_info_architecture_riscv_64" />
+  </unix:uname_test>
+  <unix:uname_object comment="64 bit architecture"
+  id="object_system_info_architecture_riscv_64" version="1" />
+  <unix:uname_state comment="64 bit architecture"
+  id="state_system_info_architecture_riscv_64" version="1">
+    <unix:processor_type operation="equals">riscv64</unix:processor_type>
+  </unix:uname_state>
+</def-group>
diff --git a/shared/macros/01-general.jinja b/shared/macros/01-general.jinja
index 80dc6eeca5..d8336855c3 100644
--- a/shared/macros/01-general.jinja
+++ b/shared/macros/01-general.jinja
@@ -1262,6 +1262,8 @@ Part of the grub2_bootloader_argument(_absent) templates.
 {{% macro grub_command(action, arg_name_value=None) -%}}
 {{%- if "update-grub" in grub_helper_executable -%}}
     {{%- set grub_helper_args = [] -%}}
+{{%- elif "grub-mkconfig" in grub_helper_executable -%}}
+    {{%- set grub_helper_args = ["-o " ~ grub2_boot_path ~ "/grub.cfg"] -%}}
 {{%- elif "grub2-mkconfig" in grub_helper_executable -%}}
     {{%- if 'suse' in families -%}}
     {{%- set grub_helper_args = ["-o " ~ grub2_boot_path ~ "/grub.cfg"] -%}}
diff --git a/shared/macros/10-ansible.jinja b/shared/macros/10-ansible.jinja
index 33417159b0..12a5053c9c 100644
--- a/shared/macros/10-ansible.jinja
+++ b/shared/macros/10-ansible.jinja
@@ -1637,6 +1637,9 @@ Part of the grub2_bootloader_argument template.
 {{% if product in ['sle12', 'sle15', 'sle16', 'slmicro5'] %}}
 - name: Update grub defaults and the bootloader menu
   ansible.builtin.command: /usr/sbin/grub2-mkconfig -o {{{ grub2_boot_path }}}/grub.cfg
+{{% elif product == 'openruyi' %}}
+- name: Update grub defaults and the bootloader menu
+  ansible.builtin.command: /usr/sbin/grub-mkconfig -o {{{ grub2_boot_path }}}/grub.cfg
 {{% elif 'debian' in product %}}
 - name: Update grub defaults and the bootloader menu
   ansible.builtin.command: /usr/sbin/update-grub
@@ -1693,6 +1696,9 @@ Part of the grub2_bootloader_argument_absent template.
 {{% if product in ['sle12', 'sle15', 'sle16'] %}}
 - name: Update grub defaults and the bootloader menu
   ansible.builtin.command: /usr/sbin/grub2-mkconfig -o {{{ grub2_boot_path }}}/grub.cfg
+{{% elif product == 'openruyi' %}}
+- name: Update grub defaults and the bootloader menu
+  ansible.builtin.command: /usr/sbin/grub-mkconfig -o {{{ grub2_boot_path }}}/grub.cfg
 {{% elif 'debian' in product %}}
 - name: Update grub defaults and the bootloader menu
   ansible.builtin.command: /usr/sbin/update-grub
diff --git a/shared/macros/10-bash.jinja b/shared/macros/10-bash.jinja
index 7ad8a71c02..c4d57cb573 100644
--- a/shared/macros/10-bash.jinja
+++ b/shared/macros/10-bash.jinja
@@ -1973,7 +1973,7 @@ Part of the grub2_bootloader_argument template.
 
 #}}
 {{% macro grub2_bootloader_argument_remediation(arg_name, arg_name_value) %}}
-{{% if 'ubuntu' in product or 'debian' in product or product in ['ol7', 'sle12', 'sle15', 'sle16', 'slmicro5'] %}}
+{{% if 'ubuntu' in product or 'debian' in product or product in ['ol7', 'sle12', 'sle15', 'sle16', 'slmicro5', 'openruyi'] %}}
 {{{ update_etc_default_grub_manually(arg_name, arg_name_value) }}}
 {{% endif -%}}
 {{{ grub_command("add", arg_name_value) }}}
@@ -1989,9 +1989,9 @@ Part of the grub2_bootloader_argument template.
 #}}
 {{%- macro update_etc_default_grub_manually_absent(arg_name) -%}}
 # Correct the form of default kernel command line in GRUB
-if grep -q -E '^GRUB_CMDLINE_LINUX=.*{{{ arg_name }}}=?.*"'  '/etc/default/grub' ; then
-       sed -i 's/\(^GRUB_CMDLINE_LINUX=".*\){{{ arg_name }}}=\?[^[:space:]]*\(.*"\)/\1 \2/' '/etc/default/grub'
-fi
+while grep -q '^GRUB_CMDLINE_LINUX=.*{{{ arg_name }}}=.*"'  '/etc/default/grub' ; do
+       sed -i 's/\(^GRUB_CMDLINE_LINUX=".*[[:space:]]\?\){{{ arg_name }}}=\?[^[:space:]]*[[:space:]]\?\(.*"\)/\1\2/' '/etc/default/grub'
+done
 {{%- endmacro %}}
 
 
@@ -2004,7 +2004,7 @@ Part of the grub2_bootloader_argument_absent template.
 
 #}}
 {{% macro grub2_bootloader_argument_absent_remediation(arg_name) %}}
-{{% if 'ubuntu' in product or product in ['ol7', 'sle12', 'sle15', 'sle16', 'slmicro5'] %}}
+{{% if 'ubuntu' in product or product in ['ol7', 'sle12', 'sle15', 'sle16', 'slmicro5', 'openruyi'] %}}
 {{{ update_etc_default_grub_manually_absent(arg_name) }}}
 {{% endif -%}}
 {{{ grub_command("remove", arg_name) }}}
diff --git a/shared/macros/10-oval.jinja b/shared/macros/10-oval.jinja
index 6466728189..8d3f2aa4e4 100644
--- a/shared/macros/10-oval.jinja
+++ b/shared/macros/10-oval.jinja
@@ -1796,7 +1796,7 @@ The macros generates the OVAL test including the dependent OVAL object and OVAL
 {{#
   Macro for checking the system architecture in /proc/sys/kernel/{osrelease,arch}
 
-  :param arch: system architecture (x86_64, aarch64, s90x, ppc64le, ...)
+  :param arch: system architecture (x86_64, aarch64, s90x, ppc64le, riscv64 ...)
   :type arch: str
 #}}
 {{%- macro oval_check_proc_sys_kernel_osrelease_arch(arch) -%}}
diff --git a/shared/templates/accounts_password/bash.template b/shared/templates/accounts_password/bash.template
index 75420036ed..38cd2a7ac3 100644
--- a/shared/templates/accounts_password/bash.template
+++ b/shared/templates/accounts_password/bash.template
@@ -6,13 +6,13 @@
 
 {{{ bash_instantiate_variables("var_password_pam_" ~ VARIABLE) }}}
 
-{{% if 'ol' in families or 'rhel' in product %}}
+{{% if 'ol' in families or 'rhel' in product or product == 'openruyi' %}}
 if grep -sq {{{ VARIABLE }}} /etc/security/pwquality.conf.d/*.conf ; then
     sed -i "/{{{ VARIABLE }}}/d" /etc/security/pwquality.conf.d/*.conf
 fi
 {{% endif %}}
 
-{{% if "ol" in families %}}
+{{% if "ol" in families or product == "openruyi" %}}
 {{{ bash_remove_pam_module_option_configuration('/etc/pam.d/system-auth',
                                   'password',
                                   '',
diff --git a/shared/templates/accounts_password/oval.template b/shared/templates/accounts_password/oval.template
index bf4b86b19b..bddde19d20 100644
--- a/shared/templates/accounts_password/oval.template
+++ b/shared/templates/accounts_password/oval.template
@@ -1,4 +1,4 @@
-{{% if 'ol' in families or 'rhel' in product %}}
+{{% if 'ol' in families or 'rhel' in product or product == 'openruyi' %}}
 {{% set filepath_regex="^/etc/security/pwquality\.conf(\.d/[^/]+\.conf)?$" %}}
 {{% else %}}
 {{% set filepath_regex="^" + pwquality_path + "$" %}}
@@ -11,7 +11,7 @@
       <criteria operator="OR">
         <criterion comment="pwquality.conf" test_ref="test_password_pam_pwquality_{{{ VARIABLE }}}" />
       </criteria>
-      {{% if "ol" in families %}}
+      {{% if "ol" in families or product == "openruyi" %}}
       <criterion comment="{{{ VARIABLE }}} is not overwritten in system-auth"
         test_ref="test_password_pam_pwquality_{{{ VARIABLE }}}_not_overwritten"/>
       <criterion comment="{{{ VARIABLE }}} is not overwritten in password-auth"
@@ -20,7 +20,7 @@
     </criteria>
   </definition>
 
-  {{% if "ol" in families %}}
+  {{% if "ol" in families or product == "openruyi" %}}
   <ind:textfilecontent54_test check="all" check_existence="none_exist"
     comment="check the configuration of /etc/pam.d/system-auth doens't override pwquality.conf"
     id="test_password_pam_pwquality_{{{ VARIABLE }}}_not_overwritten" version="1">
diff --git a/shared/templates/grub2_bootloader_argument/bash.template b/shared/templates/grub2_bootloader_argument/bash.template
index f43f3484c0..2a96387a7a 100644
--- a/shared/templates/grub2_bootloader_argument/bash.template
+++ b/shared/templates/grub2_bootloader_argument/bash.template
@@ -1,4 +1,4 @@
-# platform = multi_platform_debian,multi_platform_fedora,multi_platform_ol,multi_platform_rhel,multi_platform_rhv,multi_platform_sle,multi_platform_slmicro,multi_platform_ubuntu,multi_platform_almalinux
+# platform = multi_platform_debian,multi_platform_fedora,multi_platform_ol,multi_platform_rhel,multi_platform_rhv,multi_platform_sle,multi_platform_slmicro,multi_platform_ubuntu,multi_platform_almalinux,multi_platform_openruyi
 {{#
    See the OVAL template for more comments.
    Product-specific categorization should be synced across all template content types
diff --git a/shared/templates/grub2_bootloader_argument_absent/bash.template b/shared/templates/grub2_bootloader_argument_absent/bash.template
index 06db2fbc25..fbe6e9da15 100644
--- a/shared/templates/grub2_bootloader_argument_absent/bash.template
+++ b/shared/templates/grub2_bootloader_argument_absent/bash.template
@@ -1,4 +1,4 @@
-# platform = multi_platform_rhel,multi_platform_fedora,multi_platform_ol,multi_platform_almalinux,multi_platform_rhv,multi_platform_ubuntu,multi_platform_sle
+# platform = multi_platform_rhel,multi_platform_fedora,multi_platform_ol,multi_platform_almalinux,multi_platform_rhv,multi_platform_ubuntu,multi_platform_sle,multi_platform_openruyi
 {{#
    See the OVAL template for more comments.
    Product-specific categorization should be synced across all template content types
diff --git a/shared/templates/service_disabled/bash.template b/shared/templates/service_disabled/bash.template
index 4c93f2b84d..136020e210 100644
--- a/shared/templates/service_disabled/bash.template
+++ b/shared/templates/service_disabled/bash.template
@@ -1,4 +1,4 @@
-# platform = multi_platform_rhel,multi_platform_fedora,multi_platform_ol,multi_platform_almalinux,multi_platform_rhv,multi_platform_sle,multi_platform_slmicro,multi_platform_ubuntu
+# platform = multi_platform_rhel,multi_platform_fedora,multi_platform_ol,multi_platform_almalinux,multi_platform_rhv,multi_platform_sle,multi_platform_slmicro,multi_platform_ubuntu,multi_platform_openruyi
 # reboot = false
 # strategy = disable
 # complexity = low
diff --git a/shared/templates/service_enabled/bash.template b/shared/templates/service_enabled/bash.template
index 4e207a79d7..5877749868 100644
--- a/shared/templates/service_enabled/bash.template
+++ b/shared/templates/service_enabled/bash.template
@@ -1,4 +1,4 @@
-# platform = multi_platform_fedora,multi_platform_ol,multi_platform_rhel,multi_platform_rhv,multi_platform_sle,multi_platform_slmicro,multi_platform_ubuntu,multi_platform_almalinux,multi_platform_debian
+# platform = multi_platform_fedora,multi_platform_ol,multi_platform_rhel,multi_platform_rhv,multi_platform_sle,multi_platform_slmicro,multi_platform_ubuntu,multi_platform_almalinux,multi_platform_debian,multi_platform_openruyi
 # reboot = false
 # strategy = enable
 # complexity = low
diff --git a/shared/templates/sysctl/bash.template b/shared/templates/sysctl/bash.template
index 0f0e48edee..86f960db91 100644
--- a/shared/templates/sysctl/bash.template
+++ b/shared/templates/sysctl/bash.template
@@ -1,4 +1,4 @@
-# platform = multi_platform_debian,multi_platform_fedora,multi_platform_ol,multi_platform_almalinux,multi_platform_rhel,multi_platform_rhv,multi_platform_sle,multi_platform_slmicro,multi_platform_ubuntu
+# platform = multi_platform_debian,multi_platform_fedora,multi_platform_ol,multi_platform_almalinux,multi_platform_rhel,multi_platform_rhv,multi_platform_sle,multi_platform_slmicro,multi_platform_ubuntu,multi_platform_openruyi
 # reboot = true
 # strategy = disable
 # complexity = low
diff --git a/ssg/constants.py b/ssg/constants.py
index f104ecec2c..63bcdf287f 100644
--- a/ssg/constants.py
+++ b/ssg/constants.py
@@ -53,6 +53,7 @@ product_directories = [
     'rhcos4',
     'ol7', 'ol8', 'ol9', 'ol10',
     'openeuler2203',
+    'openruyi',
     'opensuse',
     'openembedded',
     'rhel8', 'rhel9', 'rhel10',
@@ -228,6 +229,7 @@ FULL_NAME_TO_PRODUCT_MAPPING = {
     "Oracle Linux 9": "ol9",
     "Oracle Linux 10": "ol10",
     "openEuler 2203": "openeuler2203",
+    "openRuyi": "openruyi",
     "openSUSE": "opensuse",
     "Red Hat Enterprise Linux 8": "rhel8",
     "Red Hat Enterprise Linux 9": "rhel9",
@@ -289,7 +291,7 @@ REFERENCES = dict(
 
 
 MULTI_PLATFORM_LIST = ["rhel", "fedora", "rhv", "debian", "ubuntu",
-                       "openeuler", "kylinserver",
+                       "openeuler", "openruyi", "kylinserver",
                        "opensuse", "sle", "tencentos", "ol", "ocp", "rhcos",
                        "example", "eks", "alinux", "anolis", "openembedded", "al",
                        "slmicro", "almalinux"]
@@ -304,6 +306,7 @@ MULTI_PLATFORM_MAPPING = {
     "multi_platform_fedora": ["fedora"],
     "multi_platform_kylinserver": ["kylinserver10"],
     "multi_platform_openeuler": ["openeuler2203"],
+    "multi_platform_openruyi": ["openruyi"],
     "multi_platform_opensuse": ["opensuse"],
     "multi_platform_ol": ["ol7", "ol8", "ol9", "ol10"],
     "multi_platform_ocp": ["ocp4"],
@@ -434,6 +437,7 @@ MAKEFILE_ID_TO_PRODUCT_MAP = {
     'eap': 'JBoss Enterprise Application Platform',
     'fuse': 'JBoss Fuse',
     'openeuler': 'openEuler',
+    'openruyi': 'openRuyi',
     'opensuse': 'openSUSE',
     'sle': 'SUSE Linux Enterprise',
     'slmicro': 'SUSE Linux Enterprise Micro',
diff --git a/tests/shared/grub2.sh b/tests/shared/grub2.sh
index 10df2e2ad0..4d6dbae077 100644
--- a/tests/shared/grub2.sh
+++ b/tests/shared/grub2.sh
@@ -1,4 +1,4 @@
-{{% if 'ubuntu' in product %}}
+{{% if 'ubuntu' in product or product == 'openruyi' %}}
 test -n "$GRUB_CFG_ROOT" || GRUB_CFG_ROOT=/boot/grub
 {{% else %}}
 test -n "$GRUB_CFG_ROOT" || GRUB_CFG_ROOT=/boot/grub2
@@ -7,6 +7,8 @@ test -n "$GRUB_CFG_ROOT" || GRUB_CFG_ROOT=/boot/grub2
 function set_grub_uefi_root {
 	if grep NAME /etc/os-release | grep -iq fedora; then
 		GRUB_CFG_ROOT=/boot/grub2
+	elif grep NAME /etc/os-release | grep -iq "openruyi"; then
+		GRUB_CFG_ROOT=/boot/grub
 	elif grep NAME /etc/os-release | grep -iq "Red Hat"; then
 		if grep VERSION /etc/os-release | grep -q '9\.'; then
 			GRUB_CFG_ROOT=/boot/grub2
-- 
2.34.1

